### PR TITLE
[processor/filterprocessor] Ability to filter spans

### PR DESCRIPTION
**Pull request by boostchicken**
Signed-off-by: John <dorman@overlooked.us>

**Description:** <Describe what has changed.>

Added the ability to use the filterprocessor for spans.  The documentation for filter processor currently claims it can do this, however it cannot.

Issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5838  ... plus many more
**Testing:**
Tested locally, all filtering logic is tested by the filterspan internal package, this is just a re-use of that
There are also unit tests that test the filter logic, beyond the filterspan code
**Documentation:**
Added README.md with a usage example as well as pointing to complete documentation in the attributes processor.

### CODE REVIEW

**Comment by boostchicken** [comment]
CLA pending internal Sony process, should be here shortly

**Review by mjallday — APPROVED** [review]
🙌

**Comment by boostchicken** [comment]
Resurrecting this while waitinf ro CLA

**Comment by boostchicken** [comment]
@mjallday can you remove the stale still waiting on CLA

**Comment by boostchicken** [comment]
I moved to teh CLBA to my personal accont since no code was written on Sony hardware or for SOny requirements

**Comment by mjallday** [comment]
I can’t remove the stale message. I’m just an interested neutral party and not affiliated with otel project beyond creating #5838.

**Comment by boostchicken** [comment]
@tigrannajaryan

Sorted and ready for your review / merge

**Comment by tigrannajaryan** [comment]
Span filtering was historically considered a bad feature since it can result in broken traces. On hand I understand that sometimes you really want to get rid of some noisy/irrelevant spans, but on the other hand unsuspecting users may misuse the capability.

I would like to know what other @open-telemetry/collector-contrib-approvers think about having this capability.

**Comment by jpkrohling** [comment]
Before doing an in-depth review, I have a few questions:

- Is this PR taking care of remapping the child-spans of the spans being removed?
- Would this feature require the usage of something like the groupbytrace, so that the processor has a view of the entire trace (possibly because of the point above)?

**Comment by Aneurysm9** [comment]
I share the same concerns as Tigran and Juraci.  We don't really have a good way of knowing whether a span will have children, so dropping spans can be dangerous.  I know there are situations where an operator can know that it is safe to drop spans, though, so maybe it is okay to give them this capability with a giant warning:  ***HERE BE FOOTGUNS!***

**Comment by jpkrohling** [comment]
> Is there a good alternative to dropping spans and what’s the danger?

There's some research on this topic, but nothing you can use out of the box right now. The only good solution is to tune down your instrumentation, by not creating spans for things that don't matter that much.

> we get charged for the data we send to them

Do you pay per span, or per bytes? You can try to slim down your spans by removing all tags, leaving only the timing information. But really, if you have spans as part of your trace that you'd rather drop, then you shouldn't be generating them in the first place.

**Comment by BlakeWills** [comment]
I'm not the original commenter but I've been waiting for this PR to be merged so we can start filtering out unwanted spans.

> There's some research on this topic, but nothing you can use out of the box right now. The only good solution is to tune down your instrumentation, by not creating spans for things that don't matter that much.

I wouldn't say this is a "good" solution, but rather the only solution. In our use-case, we don't actually have control over some of the unwanted spans being created. For example we're using the Azure SDKs which makes a bunch of http requests to different domains/endpoints (especially in the functions SDK) and some of these requests are significant in volume.

Currently, every time we spot spans that we'd like to exclude we need to go into our base OpenTelemetry libraries, update the http filter, and then ask each internal development team to go and update the packages in each service that they own. It's a really poor development experience.

> Do you pay per span, or per bytes?

Both. In our situation DataDog charges for GB ingest and the number of spans retained by index filters, by default it samples across all spans so the unwanted spans are included. At least for us, this means the trade off is between the potential for broken traces by configuring the collector incorrectly, or additional financial cost to the business, and DataDog really isn't cheap.

At least in my experience, the majority of spans I'd like to exclude are single span traces (sdk http invocations, status endpoint requests, etc), or traces that continue forever with something like continuous http polling. It's not like we're trying to exclude a particular span from a trace that we care about, so our potential for broken traces is minimal, but I understand that I don't speak for everyone.

In summary, this is an extremely useful feature in terms of keeping the cost of distributed tracing down, and the current "solution" leads to a really poor development experience, it's much easier to deploy a change to the collector instances than it is to multiple services.

**Comment by jpkrohling** [comment]
Thanks for the clarification, @BlakeWills. Perhaps we could start with a big warning saying that this will *only* work for leaf nodes and that if applied to any spans with children, it *will* cause traces to be broken?

**Comment by tigrannajaryan** [comment]
>At least in my experience, the majority of spans I'd like to exclude are single span traces

Can you use https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor to exclude the entire single-span trace?

**Comment by mjallday** [comment]
based on this comment it appears that the tail sampling processor may be deprecated.

https://github.com/open-telemetry/opentelemetry-collector/issues/2310#issuecomment-777179882

**Comment by boostchicken** [comment]
Hey, let me give some background on why I wrote this.  Auto instrument a few thousand redis clients while handling millions of connections per second, all the sudden you don't care for the redis spans that tell you nothing.

**Comment by jpkrohling** [comment]
> tail sampling processor may be deprecated

There's a discussion about coming up with a plan to deprecate it. It's not deprecated yet. Even if we deprecate, we'll provide a migration path supporting the same current use-cases.

**Comment by fbogsany** [comment]
There's value in filtering spans in a pipeline that includes a `spanmetrics` processor. I'm currently experimenting with this by combining the `probabilistic_sampler` processor and the `attributes` processor to filter out everything but HTTP spans ahead of the `spanmetrics` processor:
```yaml
processors:
  probabilistic_sampler: {}
  # Remove any sampling.priority set upstream to ensure we're only metricizing the spans we want.
  attributes/unsample:
    actions:
      - key: sampling.priority
        action: delete
  # Filter out everything except HTTP spans. This processor will set the sampling.priority
  # on all matching spans, then the probabilistic_sampler processor will drop anything that
  # isn't tagged by this processor.
  attributes/sample:
    include:
      match_type: strict
      attributes:
        - key: http.method
    actions:
      - key: sampling.priority
        value: 1
        action: upsert
...
service:
  pipelines:
    traces:
      receivers: [pubsub]
      processors:
        - memory_limiter
        - attributes/unsample
        - attributes/sample
        - probabilistic_sampler
        - spanmetrics
      exporters: [nop]
```

**Comment by jpkrohling** [comment]
@fbogsany, are those spans reaching a final tracing backend? If so, would you mind sharing a screenshot showing how they look without the filter and with the filter? Just removing HTTP spans would, if I'm understanding this correctly, this would mean that the trace would have orphaned children, so, can't be displayed correctly anymore.

**Comment by fbogsany** [comment]
@jpkrohling no, they're not reaching a backend. After deriving metrics from the filtered spans, I'm terminating the trace pipeline with a `nop` exporter. I have other pipelines that send complete traces to various backends.

**Comment by fbogsany** [comment]
An obvious alternative for my use-case is to add matcher support to the `spanmetrics` processor.

**Comment by jpkrohling** [comment]
@Aneurysm9, @tigrannajaryan, given the discussion and the arguments provided in this issue, do you think it's worth accepting this given enough warnings to users of this feature?

File: processor/filterprocessor/README.md
Comment (Aneurysm9) [inline processor/filterprocessor/README.md:207]: The top of this file should be updated to indicate that the processor supports traces.  This information should probably go up there as well, alongside the required parameters for logs and metrics.

Please also include a prominent warning that using this processor can result in broken traces and that it is the operator's responsibility to ensure that it only filters out leaf spans.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by Aneurysm9** [comment]
> @Aneurysm9, @tigrannajaryan, given the discussion and the arguments provided in this issue, do you think it's worth accepting this given enough warnings to users of this feature?

This needs all the warnings in all the places, but it does what it says on the tin and I'm not sure that being incredibly dangerous is enough reason to prevent the inclusion of an incredibly useful feature.

File: internal/coreinternal/processor/filterspan/filterspan.go
Comment (jpkrohling) [inline internal/coreinternal/processor/filterspan/filterspan.go:115]: Is this related to another issue?

File: processor/filterprocessor/filter_processor_traces.go
Comment (jpkrohling) [inline processor/filterprocessor/filter_processor_traces.go:44]: If you are using this only in the logs, perhaps `<none>` would be a better default?

File: processor/filterprocessor/filter_processor_traces.go
Comment (jpkrohling) [inline processor/filterprocessor/filter_processor_traces.go:115]:
```suggestion
		if !fsp.include.MatchSpan(span, resource, library) {
```

File: processor/filterprocessor/filter_processor_traces.go
Comment (jpkrohling) [inline processor/filterprocessor/filter_processor_traces.go:121]:
```suggestion
		if fsp.exclude.MatchSpan(span, resource, library) {
```

File: processor/filterprocessor/filter_processor_traces_test.go
Comment (jpkrohling) [inline processor/filterprocessor/filter_processor_traces_test.go:145]: At this point, you should probably use `require` instead of `assert`, given that you are still setting up the scene for the test.

File: processor/filterprocessor/filter_processor_traces_test.go
Comment (jpkrohling) [inline processor/filterprocessor/filter_processor_traces_test.go:157]:
```suggestion
			// If all traces got filtered you shouldn't even have ResourceSpans
```

File: processor/filterprocessor/filter_processor_traces_test.go
Comment (jpkrohling) [inline processor/filterprocessor/filter_processor_traces_test.go:159]: You can use `require.Len` instead.

Comment (jpkrohling) [reply processor/filterprocessor/filter_processor_traces_test.go:159]: And actually, this is the place to use `assert` :-)

File: processor/filterprocessor/README.md
Comment (boostchicken) [reply processor/filterprocessor/README.md:207]: I am happy to write it, but do you really think our users wouldn't take that into consideration?  I mean beyond that when we get to more advanced sampling their heads might explode :)

File: internal/coreinternal/processor/filterspan/filterspan.go
Comment (boostchicken) [reply internal/coreinternal/processor/filterspan/filterspan.go:115]: I see github lost my comment, Yes Datadog does not conform to spec so you ahve to check both places

File: processor/filterprocessor/filter_processor_traces_test.go
Comment (boostchicken) [reply processor/filterprocessor/filter_processor_traces_test.go:145]: good call, fyi this is copy and paste carry over, so this type of stuff is all over the place

**Comment by boostchicken** [comment]
@jpkrohling this is done.  Please mere? here is done A

**Comment by jpkrohling** [comment]
I just rebased and fixed the conflict with the changelog. There was a conflict in the readme file as well, would you mind checking that I didn't mess it up?

**Comment by boostchicken** [comment]
Sure thing man,  I am super curious how this went from no conflicts etc to this bad?  Nothing much has changed?

**Comment by jpkrohling** [comment]
> Sure thing man, I am super curious how this went from no conflicts etc to this bad? Nothing much has changed?

The changelog is in constant evolution and it's easy to get a conflict on it these days.

**Comment by jpkrohling** [comment]
Given @tigrannajaryan's previous comments, it would be good to get his approval as well before merging.

File: internal/coreinternal/processor/filterspan/filterspan.go
Comment (tigrannajaryan) [inline internal/coreinternal/processor/filterspan/filterspan.go:142]: `service.name` is a [Resource semantic convention](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#service). It is not supposed to be recorded as a Span attribute. Why is this check?

File: processor/filterprocessor/README.md
Comment (tigrannajaryan) [inline processor/filterprocessor/README.md:3]:
```suggestion
- Supported pipeline types: logs, metrics, traces
```

File: processor/filterprocessor/README.md
Comment (tigrannajaryan) [inline processor/filterprocessor/README.md:10]:
```suggestion
- Spans based on span names, and resource attributes, all with full regex support
```

File: processor/filterprocessor/README.md
Comment (tigrannajaryan) [inline processor/filterprocessor/README.md:209]: What does `diff` do here?

File: processor/filterprocessor/README.md
Comment (tigrannajaryan) [inline processor/filterprocessor/README.md:212]: What's "TRLY"?

File: processor/filterprocessor/README.md
Comment (tigrannajaryan) [inline processor/filterprocessor/README.md:211]: Please review the spelling or reword, I am not quite sure how to read this.

**Comment by tigrannajaryan** [comment]
Make sure the build passes.

**Comment by bogdandrutu** [comment]
@tigrannajaryan @boostchicken @anuraaga  I think this should be a simple "drop span" action in the new transform processor.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan @boostchicken @anuraaga I think this should be a simple "drop span" action in the new transform processor.

I agree. However, if we keep the filter processor then I think there is no reason for now not to allow filtering spans in filter processor as well.

**Comment by trutty** [comment]
HI, is this PR still worked on? The feature to drop/filter spans is a really useful one (wether via a filter processor or the transform-processor).

**Comment by danoliver1** [comment]
Please can this be re-opened / merged?

**Comment by TylerHelmuth** [comment]
The transform processor is now in a state where a "drop span" function could be added if anyone wants to implement it.  For this PR, I believe there are still outstanding questions from @tigrannajaryan that need handled before it could be merged.

**Comment by boostchicken** [comment]
Hey all,

My apologies on just leaving this hanging.  There were some issues out of my control precluding me from finishing it up.  Let me see if I can get this thing updated and fixed with the latest version.

**Comment by boostchicken** [comment]
@dmitryax please re-open I have addressed all your feedback and updated this

**Comment by boostchicken** [comment]
@dmitryax is check-links going to block this merge?  I didn't touch the links that are broken someone must have removed something from the repo?

File: processor/filterprocessor/filter_processor_traces.go
Comment (pmm-sumo) [inline processor/filterprocessor/filter_processor_traces.go:22]: Does this empty line pass formatting checks?

File: processor/filterprocessor/filter_processor_traces_test.go
Comment (pmm-sumo) [inline processor/filterprocessor/filter_processor_traces_test.go:23]: I think the imports should not have this blank line within a group

**Review by pmm-sumo — APPROVED** [review]
Minor formatting comments but the code looks good otherwise

File: processor/filterprocessor/filter_processor_traces_test.go
Comment (boostchicken) [reply processor/filterprocessor/filter_processor_traces_test.go:23]: This is how make fmt did it

File: processor/filterprocessor/filter_processor_traces.go
Comment (boostchicken) [reply processor/filterprocessor/filter_processor_traces.go:22]: Yes

**Review by codeboten — APPROVED** [review]

**Pull request merged**
